Libraries/LibC/spawn.cpp

```diff
@@ -0,0 +1,250 @@
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* posix_spawn and friends
+ *
+ * values from POSIX standard unix specification
+ *
+ * https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/spawn.h.html
+ */
+
+#include <spawn.h>
+
+#include <assert.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+extern "C" {
+
+[[noreturn]] static void posix_spawn_child(
+    const char* RESTRICT path,
+    const posix_spawn_file_actions_t* RESTRICT file_actions,
+    const posix_spawnattr_t* RESTRICT attr,
+    char* const argv[RESTRICT],
+    char* const envp[RESTRICT],
+    bool search_path)
+{
+    if (file_actions) {
+        // FIXME
+    }
+    if (attr) {
+        short flags = attr->flags;
+        if (flags & POSIX_SPAWN_SETPGROUP)
+            if (setpgid(0, attr->pgroup) < 0)
+                exit(127);
+        if (flags & POSIX_SPAWN_SETSCHEDPARAM)
+            if (sched_setparam(0, &attr->schedparam) < 0)
+                exit(127);
+        if (flags & POSIX_SPAWN_SETSCHEDULER)
+            if (sched_setparam(0, &attr->schedparam) < 0)
+                exit(127);
+        if (flags & POSIX_SPAWN_SETSIGDEF) {
+            struct sigaction default_action;
+            default_action.sa_flags = 0;
+            sigemptyset(&default_action.sa_mask);
+            default_action.sa_handler = SIG_DFL;
+
+            sigset_t sigdefault = attr->sigdefault;
+            for (int i = 0; i < NSIG; ++i)
+                if (sigismember(&sigdefault, i) && sigaction(i, &default_action, nullptr) < 0)
+                    exit(127);
+        }
+        if (flags & POSIX_SPAWN_SETSIGMASK)
+            if (sigprocmask(SIG_SETMASK, &attr->sigmask, nullptr) != 0)
+                exit(127);
+        // FIXME: POSIX_SPAWN_RESETIDS, POSIX_SPAWN_SETSCHEDULER
+    }
+
+    if (search_path)
+        execvpe(path, argv, envp);
+    execve(path, argv, envp);
+    __assertion_failed("returned from exec");
```

Comment:
* We don't call `__assertion_failed()` directly, we use `ASSERT_NOT_REACHED()` or `ASSERT(something)`
* In this case, exec clearly can return due to an error, so just asserting like that is not appropriate. Perhaps something like this:
   ```cpp
   perror("exec");
   exit(1);
   ```
    would be more appropriate. It'd be even better to include the path we attempted to exec in the error message.

Reply:
Good point, thanks. Added a few perror()s for now.

I also looked over the style guide again and saw the (good) suggestion to prefer enums over boolean parameters so I changed that, but then I figured I can just pass in the right exec directly. Let me know if you think that's overly cutesy, then I'll make it an enum insteaad.